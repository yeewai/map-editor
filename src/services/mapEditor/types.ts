export interface State {
    zoom: number,
    panV: number,
    panH: number
}
