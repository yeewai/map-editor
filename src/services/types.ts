import { State as structureDefinitionType } from './structureDefinitions/types';
import { State as worldType } from './worlds/types';
import { State as mapEditorType } from './mapEditor/types';

import { defaultState as structureDefinitionState } from './structureDefinitions/reducer';
import { defaultState as worldState } from './worlds/reducer';
import { defaultState as mapEditorState } from './mapEditor/reducer';

export interface StateTree {
    structureDefinitions: structureDefinitionType,
    worlds: worldType,
    mapEditor: mapEditorType
}

export const defaultStateTree: StateTree = {
    structureDefinitions: structureDefinitionState,
    worlds: worldState,
    mapEditor: mapEditorState
}

export interface Action {
    type: string,
    payload?: any
}

export const defaultAction:Action = {
    type: ""
}
