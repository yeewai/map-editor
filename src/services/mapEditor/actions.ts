import { ActionCreator } from 'redux';
import { Action } from 'services/types';

// -------------------------------------------------------------
// "Camera" controls
// -------------------------------------------------------------
export const zoomIn: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMIN'
});

export const zoomOut: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMOUT'
});

export const zoomReset: ActionCreator<Action> = () => ({
    type: 'mapEditor/ZOOMRESET'
});

export const panLeft: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANLEFT'
});

export const panRight: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANRIGHT'
});

export const panUp: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANUP'
});

export const panDown: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANDOWN'
});

export const panReset: ActionCreator<Action> = () => ({
    type: 'mapEditor/PANRESET'
});
