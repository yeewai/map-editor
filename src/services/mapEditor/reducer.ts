import { Action } from 'services/types';
import * as Types from './types';

export const defaultState: Types.State = {
    zoom: 1,
    panV: 0,
    panH: 0
}

export const ZOOMRATE = 1.5;
export const MAXZOOM = 10;
export const PANRATE = 1;

const mapEditorReducer = (state = defaultState, action: Action ): Types.State => {
    switch(action.type) {
        case 'mapEditor/ZOOMIN':
            return { ...state, zoom: Math.min(state.zoom * ZOOMRATE, MAXZOOM) };
        case 'mapEditor/ZOOMOUT':
            return { ...state, zoom: state.zoom / ZOOMRATE };
        case 'mapEditor/ZOOMRESET':
            return { ...state, zoom: defaultState.zoom };
        case 'mapEditor/PANUP':
            return {...state, panV: state.panV - PANRATE}
        case 'mapEditor/PANDOWN':
            return {...state, panV: state.panV + PANRATE}
        case 'mapEditor/PANLEFT':
            return {...state, panH: state.panH - PANRATE}
        case 'mapEditor/PANRIGHT':
            return {...state, panH: state.panH + PANRATE}
        case 'mapEditor/PANRESET':
            return {...state, panH: defaultState.panH, panV: defaultState.panV}
        default:
            return state;
    }
}


export default mapEditorReducer;
