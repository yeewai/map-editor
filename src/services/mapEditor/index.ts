import reducer from "./reducer";
import * as mapEditorActions from "./actions";
import * as mapEditorSelectors from "./selectors";
import * as mapEditorTypes from "./types";

export { mapEditorActions, mapEditorSelectors, mapEditorTypes };
export default reducer;
