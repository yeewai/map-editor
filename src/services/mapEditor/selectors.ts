import { StateTree } from 'services/types';

export const getZoom = ( state: StateTree ) => state.mapEditor.zoom;
export const getPan = (state: StateTree ) => ({ h: state.mapEditor.panH, v: state.mapEditor.panV });
