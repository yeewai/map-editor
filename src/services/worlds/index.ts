import reducer from "./reducer";
import * as worldActions from "./actions";
import * as worldSelectors from "./selectors";
import * as worldTypes from "./types";

export { worldActions, worldSelectors, worldTypes };
export default reducer;
