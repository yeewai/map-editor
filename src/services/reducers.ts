import { routerReducer } from 'react-router-redux';
import { combineReducers } from 'redux';
import { modalReducer } from 'ever-modal';
import { reducer as formReducer } from 'redux-form'

import { StateTree } from './types';
import structureDefinitions from './structureDefinitions';
import worlds from './worlds';

const rootReducer = combineReducers<StateTree>({
    router: routerReducer,
    modals: modalReducer,
    form: formReducer,
    structureDefinitions,
    worlds

});

export default rootReducer;
