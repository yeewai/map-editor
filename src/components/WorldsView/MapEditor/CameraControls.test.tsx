import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { CameraControls, mapDispatchToProps } from './CameraControls';
import { mapEditorActions } from 'services/mapEditor';

Enzyme.configure({ adapter: new Adapter() });

describe("Camera Controls for Map", () => {
    let state, sandbox, store, wrapper, dispatchSpy;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
        state = {};
        dispatchSpy = sandbox.spy();
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders camera controls", () => {
		wrapper = shallow(<CameraControls
            zoomIn={sinon.spy()} zoomOut={sinon.spy()} zoomReset={sinon.spy()}
            panUp={sinon.spy()} panDown={sinon.spy()} panLeft={sinon.spy()} panRight={sinon.spy()} panReset={sinon.spy()}
        />)
		expect(wrapper).toMatchSnapshot();
	});

    it ("handles all the camera zoom and pan actions", () => {
        const actions = ["zoomIn", "zoomOut", "zoomReset", "panUp", "panDown", "panLeft", "panRight", "panReset"];

        actions.forEach( a => {

            const spy = sandbox.stub(mapEditorActions, a);

            mapDispatchToProps(dispatchSpy)[a]();
            expect(spy.called).toBe(true);

            sandbox.restore();
        })

    })

});
