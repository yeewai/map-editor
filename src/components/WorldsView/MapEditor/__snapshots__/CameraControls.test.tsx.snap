// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera Controls for Map renders camera controls 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Camera Controls
</h3>,
        <Button
          color="secondary"
          onClick={undefined}
          tag="button"
>
          Zoom In 
</Button>,
        <Button
          color="secondary"
          onClick={undefined}
          tag="button"
>
          Zoom Out 
</Button>,
        <Button
          color="secondary"
          onClick={undefined}
          tag="button"
>
          Reset Zoom
</Button>,
        <Button
          color="info"
          onClick={undefined}
          tag="button"
>
          Pan Up
</Button>,
        <Button
          color="info"
          onClick={undefined}
          tag="button"
>
          Pan Down
</Button>,
        <Button
          color="info"
          onClick={undefined}
          tag="button"
>
          Pan Left
</Button>,
        <Button
          color="info"
          onClick={undefined}
          tag="button"
>
          Pan Right
</Button>,
        <Button
          color="info"
          onClick={undefined}
          tag="button"
>
          Pan Reset
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Camera Controls",
        },
        "ref": null,
        "rendered": "Camera Controls",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Zoom In ",
          "color": "secondary",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Zoom In ",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Zoom Out ",
          "color": "secondary",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Zoom Out ",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Reset Zoom",
          "color": "secondary",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Reset Zoom",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Pan Up",
          "color": "info",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Pan Up",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Pan Down",
          "color": "info",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Pan Down",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Pan Left",
          "color": "info",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Pan Left",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Pan Right",
          "color": "info",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Pan Right",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Pan Reset",
          "color": "info",
          "onClick": undefined,
          "tag": "button",
        },
        "ref": null,
        "rendered": "Pan Reset",
        "type": [Function],
      },
    ],
    "type": "aside",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Camera Controls
</h3>,
          <Button
            color="secondary"
            onClick={undefined}
            tag="button"
>
            Zoom In 
</Button>,
          <Button
            color="secondary"
            onClick={undefined}
            tag="button"
>
            Zoom Out 
</Button>,
          <Button
            color="secondary"
            onClick={undefined}
            tag="button"
>
            Reset Zoom
</Button>,
          <Button
            color="info"
            onClick={undefined}
            tag="button"
>
            Pan Up
</Button>,
          <Button
            color="info"
            onClick={undefined}
            tag="button"
>
            Pan Down
</Button>,
          <Button
            color="info"
            onClick={undefined}
            tag="button"
>
            Pan Left
</Button>,
          <Button
            color="info"
            onClick={undefined}
            tag="button"
>
            Pan Right
</Button>,
          <Button
            color="info"
            onClick={undefined}
            tag="button"
>
            Pan Reset
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Camera Controls",
          },
          "ref": null,
          "rendered": "Camera Controls",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Zoom In ",
            "color": "secondary",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Zoom In ",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Zoom Out ",
            "color": "secondary",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Zoom Out ",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Reset Zoom",
            "color": "secondary",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Reset Zoom",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Pan Up",
            "color": "info",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Pan Up",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Pan Down",
            "color": "info",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Pan Down",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Pan Left",
            "color": "info",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Pan Left",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Pan Right",
            "color": "info",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Pan Right",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Pan Reset",
            "color": "info",
            "onClick": undefined,
            "tag": "button",
          },
          "ref": null,
          "rendered": "Pan Reset",
          "type": [Function],
        },
      ],
      "type": "aside",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
